lib/kubernetes-deploy/kubernetes_resource.rb

```diff
@@ -54,19 +41,28 @@ def timeout
       self.class.timeout
     end
 
-    def initialize(name:, namespace:, context:, file:, logger:)
+    def initialize(namespace:, context:, template:, logger:)
       # subclasses must also set these if they define their own initializer
-      @name = name
+      @name = template.fetch("metadata", {})["name"]
```

Comment:
The validation happens _immediately_ after this (unless this is a child resource, in which case it is inherently valid), so it shouldn't be a problem. Making it safer is a good idea, but we shouldn't be filling in missing names. I guess empty string would be slightly better than the current `nil` from that perspective?